src/main/java/com/google/devtools/build/lib/remote/RemoteOutputChecker.java

```diff
@@ -281,14 +281,26 @@ public boolean shouldDownloadOutput(ActionInput output, FileArtifactValue metada
     checkState(
         !(output instanceof Artifact && ((Artifact) output).isTreeArtifact()),
         "shouldDownloadOutput should not be called on a tree artifact");
-    return metadata.isRemote() && shouldDownloadOutput(output.getExecPath());
+    return metadata.isRemote()
+        && shouldDownloadOutput(
+            output.getExecPath(),
+            output instanceof Artifact artifact && artifact.isTreeArtifact()
+                ? artifact.getParent().getExecPath()
+                : null);
```

Comment:
The condition here should be `output instanceof TreeFileArtifact`. I'll fix it on import.